nixos/modules/services/networking/wireguard.nix

```diff
@@ -231,17 +237,18 @@ let
 
   generatePeerUnit = { interfaceName, interfaceCfg, peer }:
     let
+      name = if peer.name != null then peer.name else peer.publicKey;
       keyToUnitName = replaceChars
         [ "/" "-"    " "     "+"     "="      ]
         [ "-" "\\x2d" "\\x20" "\\x2b" "\\x3d" ];
-      unitName = keyToUnitName peer.publicKey;
+      unitName = keyToUnitName name;
```

Comment:
What happens if two peers are given the same name?

Reply:
Testing the PR, it looks like there is no warning or error; the first peer listed with the same name "wins" and gets put into the systemd unit.

Reply:
I feel like this calls for a change of the current peers data structure, from "list of sets" to "set of sets", e.g.:
```
networking.wireguard.interfaces = {
  business = {
    # […]
    peers = {
      fileserver = { publicKey = "…"; allowedIPs = ["…"]; };
      webserver = { publicKey = "…"; allowedIPs = ["…"]; };
    };
  };
};
```

This would be a breaking change for existing configurations, but it seems to me like the only sane way to properly handle the `name` collision issue based on the current implementation.

On the other hand, instead of having individual service units per peer, the whole `networking.wireguard` implementation could also simply make use of systemd's native WireGuard support in systemd-networkd, so we'd end up with far less complexity and implementation overhead.
As [since systemd 243](https://github.com/systemd/systemd/blob/v243/NEWS#L609), systemd-networkd also supports [`PrivateKeyFile=`](https://www.freedesktop.org/software/systemd/man/systemd.netdev.html#PrivateKeyFile=), this would also allow to take the requirement to keep the private key out of the configuration into account.

And as systemd-networkd simply provides a [`Description=` attribute for each `.netdev`](https://www.freedesktop.org/software/systemd/man/systemd.netdev.html#Description=), the whole thing would work just fine with the existing data structure and potential `name` collisions, as it wouldn't be used as primary key anymore.